gemrb/core/Game.cpp

```diff
@@ -1737,44 +1756,100 @@ void Game::CastOnRest()
-		while (specialCount--) {
-			if (special_spells[specialCount].flags & SP_REST) {
-				if (injurees.empty()) injurees = PCs; // enable some nonhealing magic too
-				// check each party member for the spell
-				while (ps--) {
-					Actor *tar = GetPC(ps, true);
-					while (tar && tar->spellbook.HaveSpell(special_spells[specialCount].resref, 0)) {
-						// we have the spell, so cast it on most injured
-						// NOTE: no distinction is made about the potency of spells
-						Actor *injuree = injurees[0];
-						tar->DirectlyCastSpell(injuree, special_spells[specialCount].resref, 0, 1, true);
-						if (injuree->GetStat(IE_HITPOINTS) == injuree->GetStat(IE_MAXHITPOINTS)) {
-							if (!injurees.empty()) {
-								injurees.erase(injurees.begin());
+		if (specialCount != -1 ) {
+			std::sort(wholeparty.begin(),wholeparty.end());
+			specialCount = core->GetSpecialSpellsCount();
+			std::multimap<ieWord,HealingResource> healingspells;
+			while (specialCount--) {
+				// Cast multi-target healing spells
+				if (special_spells[specialCount].flags & SP_HEAL_ALL) {
+					while (ps--) {
+						Actor *tar = GetPC(ps, true);
+						while (tar && tar->spellbook.HaveSpell(special_spells[specialCount].resref, 0)) {
+							tar->DirectlyCastSpell(tar, special_spells[specialCount].resref, 0, 1, true);
+							for (std::vector<Injuree>::iterator injuree=wholeparty.begin();injuree != wholeparty.end() ; ++injuree) {
+									injuree->hpneeded-=special_spells[specialCount].amount;
 							}
 						}
 					}
+					ps=ps2;
 				}
-				ps = ps2;
+				// Gather the regular healing spells
+				else if (special_spells[specialCount].flags & SP_REST) {
+					while (ps--) {
+						Actor *tar = GetPC(ps, true);
+						if (tar && tar->spellbook.HaveSpell(special_spells[specialCount].resref, 0)) {
+							HealingResource resource;
+							resource.caster=tar;
+							CopyResRef(resource.resref,special_spells[specialCount].resref);
+							resource.amount=tar->spellbook.CountSpells(special_spells[specialCount].resref,0,0);
+							healingspells.insert(std::pair<ieWord,HealingResource>(special_spells[specialCount].amount,resource));
+						}
+					}
+					ps=ps2;
+				}
+			}
+			std::sort(wholeparty.begin(),wholeparty.end());
+			// Heal who's still injured
+			while (!healingspells.empty() && wholeparty.back().hpneeded > 0) {
+				std::multimap<ieWord,HealingResource>::iterator spell=healingspells.end();
+				spell--;
+				spell->second.caster->DirectlyCastSpell(wholeparty.back().character, spell->second.resref, 0, 1, true);
+				wholeparty.back().hpneeded-=spell->first;
+				std::sort(wholeparty.begin(),wholeparty.end());
+				spell->second.amount--;
+				if (spell->second.amount == 0) {
+					healingspells.erase(spell);
+				}
+			// Here is the place to write the loop for non hp healing spells.
 			}
 		}
 	}
 }
 
+//returns a vector of character with requested state. Container will be empty if none
+bool Game::FindPCs(std::vector<Actor*>& characters, ieDword state)
```

Comment:
sure, some kind of function would be potentially useful (this isnt exactly what I had in mind), but if its not being used anywhere... plus this is rather incomplete since it would only get injured PCs. the function I had in mind was much more robust. go ahead and remove any code that isnt used by the patch.